pkg/ingester/ingester_test.go

```diff
@@ -2960,10 +3015,227 @@ func TestIngester_QueryStreamManySamplesChunks(t *testing.T) {
+	})
+
+	// Wait until it's healthy.
+	test.Poll(t, 1*time.Second, 1, func() interface{} {
+		return i.lifecycler.HealthyInstancesCount()
+	})
+
+	// Push series, alternating between different series that have chunks that encode to different sizes.
+	smallSampleSet := generateSamples(10000)   // 10k samples in chunks use about 15.4 KiB, so 64 series will take ~1 MiB
+	mediumSampleSet := generateSamples(600000) // 600k samples in chunks use about 924 KiB
+	largeSampleSet := generateSamples(1000000) // 1M samples in chunks use about 1.51 MiB
+	expectedSeries := []labels.Labels{}
+
+	for idx := 0; idx < 201; idx++ {
+		samples := smallSampleSet
+
+		if idx%80 == 0 {
+			samples = mediumSampleSet
+		} else if idx == 105 {
+			samples = largeSampleSet
+		}
+
+		l := labels.FromStrings(labels.MetricName, "foo", "l", fmt.Sprintf("%3v", idx))
+		expectedSeries = append(expectedSeries, l.Copy())
+
+		_, err = i.Push(ctx, writeRequestSingleSeries(l, samples))
+		require.NoError(t, err)
+	}
+
+	// Create a GRPC server used to query back the data.
+	serv := grpc.NewServer(grpc.StreamInterceptor(middleware.StreamServerUserHeaderInterceptor))
+	t.Cleanup(serv.GracefulStop)
+	client.RegisterIngesterServer(serv, i)
+
+	listener, err := net.Listen("tcp", "localhost:0")
+	require.NoError(t, err)
+
+	go func() {
+		require.NoError(t, serv.Serve(listener))
+	}()
+
+	c, err := client.MakeIngesterClient(listener.Addr().String(), defaultClientTestConfig())
+	require.NoError(t, err)
+	t.Cleanup(func() { c.Close() }) //nolint:errcheck
+
+	s, err := c.QueryStream(ctx, &client.QueryRequest{
+		StartTimestampMs:         0,
+		EndTimestampMs:           1000001,
+		StreamingChunksBatchSize: 64,
+
+		Matchers: []*client.LabelMatcher{{
+			Type:  client.EQUAL,
+			Name:  model.MetricNameLabel,
+			Value: "foo",
+		}},
+	})
+	require.NoError(t, err)
+
+	actualSeries := []labels.Labels{}
+	seriesBatchCount := 0
+
+	for {
+		resp, err := s.Recv()
+		require.NoError(t, err)
+
+		seriesBatchCount++
+		require.LessOrEqual(t, seriesBatchCount, 2, "expecting no more than two batches")
+
+		switch seriesBatchCount {
+		case 1:
+			require.Len(t, resp.Series, 128, "first batch should be maximum batch size")
+		case 2:
+			require.Len(t, resp.Series, 201-128, "second batch should contain remaining series")
+		}
+
+		for _, series := range resp.Series {
+			actualSeries = append(actualSeries, mimirpb.FromLabelAdaptersToLabels(series.Labels))
+		}
+
+		if resp.IsEndOfSeriesStream {
+			break
+		}
+	}
+
+	require.Equal(t, expectedSeries, actualSeries)
+
+	seriesReceivedCount := 0
+	actualSeriesPerChunksMessage := []int{}
+
+	// There are two limits when creating messages. As soon as either limit is reached for the batch, the batch is sent:
+	// 1. no message can be more than 1 MiB (unless it contains a single series whose chunks total more than 1 MiB)
+	// 2. no message can contain more series than the requested batch size
+	expectedSeriesPerChunksMessage := []int{
```

Comment:
I am worried this unit test is too sensitive to any changes that affect how we calculate the limits in streaming. While I don't think anything is going to change there drastically anytime soon, do you think there is a way that won't require manual calculation of these numbers? I don't have any immediate suggestions, so you can add a TODO as well to improve this later.

Reply:
I was thinking about this as I wrote it, and I couldn't think of anything that wasn't just re-implementing the logic we're trying to test as part of the test. Open to ideas though :) 

I think having to modify the test to reflect new behaviour is OK.